srk/src/chc.ml

```diff
@@ -91,27 +101,73 @@ module Fp = struct
       fp.rules
 
   let get_relations_declared fp =
-    BatList.of_enum (0 -- ((DynArray.length fp.rel_ctx) - 1)) 
+    BatList.of_enum (0 -- ((DynArray.length fp.rel_ctx) - 1))
 
-  let is_linear fp =
+  let is_linear rules =
     BatList.fold_left 
       (fun acc (hypo_atoms, _, _) -> acc && (List.length hypo_atoms) <= 1)
       true
-      fp.rules
+      rules
+
+  let mk_eq_by_sort srk s1 s2 =
+    assert (typ_symbol srk s1 = typ_symbol srk s2);
+    match typ_symbol_fo srk s1 with
+    | `TyInt | `TyReal -> mk_eq srk (mk_const srk s1) (mk_const srk s2)
+    | `TyBool -> mk_iff srk (mk_const srk s1) (mk_const srk s2)
+
+  let normalize srk fp =
```

Comment:
This is reasonable. My intention here is that the client can have the chc be in whatever form they want, but there are some functions (like check/solve) that if they intend to call them, they ought to normalize first.

This normalization procedure does two things (1) it ensures that each constraint has no free vars other than those that occur as a relation argument (2) it ensures that each symbol is used at most once-per-rule as an argument to a relation. I do see value in chc without these constraints.

W.r.t. (1): it is much easier for me to compute the partitioning of the chc array variables for the pmfa offset partitioning if I do so without this restriction.

W.r.t. (2): I figure that allowing this is freedom when typing up chc is just to have. It also has the added bonus that the chc generated by parse_z3fp look ever so slightly more like the original smt2 format chc.

______

If we want to remove/hide normalization likely I will 
(1) add normalization invariant (2) listed above as an invariant to all fp via add_rule, as you suggest.
(2) Not require normalization invariant (1) as an invariant to all fp, but instead whenever check/solve is entered, create a new temporary fp that has no free vars other than those that occur as a relation argument and run the check/solve on this temporary fp.

Reply:
How is it easier to compute the partition *without* the restriction?  The inputs to your partitioning algorithm are without normalization are a superset of the un-normalized inputs, no?

More broadly -- it seems there are (potentially) two different purposes here: is this module intended to be a representation of the syntax of a system of CHCs, or is it supposed to be a solver interface?  If it's the latter, the only thing that should be exposed are operations for constructing and solving CHCs.  If the solver requires normalization, everything should be normalized (transparently from the user's perspective).  If it's the former, a different set of operations ought to be exposed (and "normalize" may indeed be one of them).  If you want both, perhaps there should be two modules instead of one.

Reply:
To answer the second question:
The answer is both. I agree splitting probably makes sense here. Perhaps a module "Fp" that lives in "Chc.ml" which has submodules "Syntax" "Solve" and "SrkZ3". Potentially going a bit crazy with the nesting of modules here though.

The "normalization" function will then live in the "Solve" module (I can also just hide the procedure from the user and just run the procedure whenever solve/check is called).

______

To answer your first question: I was under the impression that you are suggesting the CHC module ensures that every CHC that is created fulfills these two requirements. Yes the offset analysis deals with normalized and un-normalized inputs, but the offset analysis proceeds by first normalizing chc in a way different than the normal form above (and most importantly not compatible with the one above). The offset-normal form has the same invariant (2) but invariant (1) is that there are no existential quants; every existential quant is skolemized.  Requiring that all chc have this normal form would mean that I cannot modify chc to be of the offset-normal form in the offset analysis and it makes the offset analysis more complicated to implement.

Reply:
You don't need an Fp module to contain everything --- everything is already contained in the file-level module Chc.

The normalization function should live in syntax, because it manipulates syntax.  The Solver module should contain logic for solving systems of CHCs -- it should call normalize in the syntax module if it needs to.   Since you want to have a module for representing syntax, the syntax should match expectations for CHCs.  The only constant symbols I would expect to see are the relation symbols -- everything else out to be a variable, implicitly universally quantified.  The decision to use constants everywhere (and to have a new type for relations rather than re-using symbols) was based on arrays not being first-order types.  If these details are hidden from clients of the module then it's not much of an issue, but if the syntax is accessible from outside this module then more care needs to be taken to make the syntax make sense.

Reply:
Hmm. I forgot why relations were a different type... maybe it makes since to switch them back to symbols now that arrays as FO types. Though this version is perhaps more future proof since it can easily be changed to handle function types (note that support for this is currently off).

_______

The current syntax of a rule is:
hypothesis: a list of relation atoms
constraint: a formula with no fv.
conclusion: a relation atom.

Let syms_constr be the list of constant symbols that occur in constraint but are not an argument to any relation atom. Let syms_atoms be the list of constant symbols that occur as an argument to a relation atom.

We interpret a rule as the following formula: \forall syms_atoms . ((\bigand_{hypo in hypothesis} hypo) /\ exists syms_constr. constraint)) => conclusion)

Perhaps this syntax isn't super intuitive - I can see why one would expect I use free_vars rather than symbols. My decision to use constant symbols rather than free_vars was that I just find constant symbols in general to be easier to work with.



______

If I make the switch from relations as their own types to symbols with func types, there is also a question of if I really even ought to have a special type for rule. Z3 encodes a rule as a formula. We definitely could just encode a rule as a formula to and throw some error if the formula doesn't look like a chc should. On one hand, I'm inclined to move closer to Z3s encoding because probably they put a lot of time thinking into their design decisions. But on the other hand I like not having the parse the rule each time to pull out the relation atoms in the formula.

Reply:
Z3's design sacrifices type safety for convenience -- (nearly) everything is of type expression.  It's a reasonable compromise.  Providing type safety and convenient accessors by having a specialized type for horn clauses is also reasonable.

One could have a system of constraints like:
```
(declare-const k Int)
(declare-fun I (Int) Bool)
(assert (I 0))
(assert (forall ((x Int) (y Int)) (=> (and (I x) (< x 10) (= y (+ x k))) (I y))))
(assert (not (I 11)))
(check-sat)
(get-model)
```
Z3 happily computes a model for this:
```
(model 
  (define-fun k () Int
    (- 4))
  (define-fun I ((x!0 Int)) Bool
    (or (not (<= (- 2) x!0)) (and (<= (- 2) x!0) (<= 0 x!0) (not (<= 2 x!0)))))
)
```
My understanding of your representation of this system changes its semantics -- k is a constant symbol that doesn't appear in a relation atom, and so it's treated as an existentially quantified variable (in a negative position, so effectively universally quantified). Using variables would avoid this, and clarify the intended interpretation of the clause to the client.  This issue also comes up in transition formulas -- this is solved (inelegantly) by attaching an "exists" predicate to the transition formula that indicates which constant symbols are to be interpreted as constants and which as skolem constants.  This has been a source of confusion, so ideally I'd like to avoid it.

Is there something that's particularly difficult about manipulating variables?  Can we make it easier?

Reply:
I see. Your understanding is correct - the scope of a symbol is in effect limited to a single rule. It seems to me that what you're proposing is equivalent to adding support for existential horn clauses, which is perhaps something we want to do, but also something I recall us for whatever reason actively not pursuing during the initial design of CHC.ml.

My annoyance with manipulating variables is that I am in love with Hashtbls (feels wrong to admit I love imperative constructs) but it doesn't really make great sense to use Hashtbls for debruijn indices because a single variable may have a different index depending on where it occurs in the formula. To give a concrete example, look at the "offset_partitioning" function in https://github.com/JakeSilverman/duet/blob/pmfa/srk/src/pmfa.ml. I have some Hashtbl from variable symbol to equivalence class, and when I want to merge variable the equiv classes of symbol a and b I just look up their cell in the Hashtbl and merge the result. The debruijn index/more functional friendly approach would be to maintain a separate map for each expression and then merge the maps (probably still using my imperative BatUrefs), but I love Hashtbls and I don't want to give them up.

Reply:
I don't mind if we don't support constants, but we shouldn't produce an incorrect result if we happen to run on a system that uses them.  This may introduce subtle bugs. For example, some iteration operators make use of symbols that are not to be treated as existentials (for example, the symbols might be represent exponential or logarithm functions, operational calculus operators, etc).  If the superlinear solver solves a linear system at some stratum and substitutes the solution for a relation into the constraint at the next stratum, are those symbols now existentially quantified?

There's nothing wrong with hash tables.  Offset partitioning is going to be imperative anyway due to the union-find.  To my eye, it would be simpler to avoid them in your offset_partitioning example: first you skolemize so that each variable reference is replaced by a unique symbol, then you use a global hash table to associate the unique symbols with union-find cells.  You could just cut out the middle man and associate variables with union-find cells, in essentially the same way that you're implementing "skolemize".  (By the way: you can implement maps over de Bruijn indicies is with [environments](https://github.com/zkincaid/duet/blob/modern/srk/src/syntax.mli#L14), which offer constant-time push and log-time find).  That is, you traverse the formula from the top down, and push a fresh union-find cell to the environment for each quantifier.  Every time you see a select, say
 (select (var 3) (var 0)), find the cell for var 3 and var 0 in the environment and merge them.  (Conceptually, var 0 shouldn't really be associated with a cell since it's of sort int.  But this is an easy trick to implement "all array symbols below a universal are merged together").

Reply:
Yes. If we have the CHC

Rule 1: x = y => R1()

Rule 2: R1() /\ x !=y => R2()

This is interpreted as 

Rule 1: (\exists x, y. x = y) => R1()

Rule2: \exists x, y. (R1() /\ x !=y) => R2()

And thus the solution to R2 is  \exists x, y, x', y'.  x' = y' /\ x !=y

This was by design (that is, I don't consider this to be a bug). But that is because in the initial design we didn't support constant symbols. 

I am happy to change this design. I'm kind of wavering between the two different design choices
Choice 1:
(1) the relation atoms have as parameters free variables
(2) every variable symbol that occurs in a rule is a constant
(3) every free variable is implicitly universally quantified in the positive position

versus choice 2:
(1) the relation atoms have as parameters variable symbols
(2) every variable symbol that occurs in a rule that is not an argument to one of the relation atoms in this rule is a constant symbol
(3) the constraint may have additional free variables and these are implicitly existentially quantified (in the negative position)

Probably going to go with the first one since it seems cleaner (but the second one will require less changes to current code).